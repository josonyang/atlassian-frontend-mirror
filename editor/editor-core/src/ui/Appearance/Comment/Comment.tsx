/* eslint-disable @atlaskit/design-system/ensure-design-token-usage/preview */
/** @jsx jsx */
import React, {
  Fragment,
  useCallback,
  useEffect,
  useMemo,
  useState,
} from 'react';

import { css, jsx } from '@emotion/react';
import classnames from 'classnames';
import type { WrappedComponentProps } from 'react-intl-next';
import { injectIntl } from 'react-intl-next';

import ButtonGroup from '@atlaskit/button/button-group';
import Button from '@atlaskit/button/custom-theme-button';
import { useSharedPluginState } from '@atlaskit/editor-common/hooks';
import { GRID_GUTTER } from '@atlaskit/editor-common/styles';
import type { OptionalPlugin } from '@atlaskit/editor-common/types';
import { WidthConsumer } from '@atlaskit/editor-common/ui';
import { ToolbarArrowKeyNavigationProvider } from '@atlaskit/editor-common/ui-menu';
import type {
  MaxContentSizePlugin,
  MaxContentSizePluginState,
} from '@atlaskit/editor-plugins/max-content-size';
import type { MediaPlugin } from '@atlaskit/editor-plugins/media';
import type { MediaPluginState } from '@atlaskit/editor-plugins/media/types';
import { tableCommentEditorStyles } from '@atlaskit/editor-plugins/table/ui/common-styles';
import { akEditorMobileBreakoutPoint } from '@atlaskit/editor-shared-styles';
import { getBooleanFF } from '@atlaskit/platform-feature-flags';
import { N40 } from '@atlaskit/theme/colors';
import { borderRadius } from '@atlaskit/theme/constants';
import { token } from '@atlaskit/tokens';

import messages from '../../../messages';
import { usePresetContext } from '../../../presets/context';
import type {
  EditorAppearance,
  EditorAppearanceComponentProps,
} from '../../../types';
import { ClickAreaBlock } from '../../Addon';
import { createEditorContentStyle } from '../../ContentStyles';
import PluginSlot from '../../PluginSlot';
import Toolbar from '../../Toolbar';
import WithFlash from '../../WithFlash';

import { MainToolbar, mainToolbarCustomComponentsSlotStyle } from './Toolbar';

const CommentEditorMargin = 14;

const commentEditorStyle = css({
  display: 'flex',
  flexDirection: 'column',
  '.less-margin .ProseMirror': {
    margin: `${token('space.150', '12px')} ${token('space.100', '8px')} ${token(
      'space.100',
      '8px',
    )}`,
  },
  minWidth: '272px',
  height: 'auto',
  backgroundColor: token('color.background.input', 'white'),
  border: `1px solid ${token('color.border', N40)}`,
  boxSizing: 'border-box',
  borderRadius: `${borderRadius()}px`,
  maxWidth: 'inherit',
  wordWrap: 'break-word',
});

const ContentArea = createEditorContentStyle(
  css(
    {
      flexGrow: 1,
      overflowX: getBooleanFF('platform.editor.table-sticky-scrollbar')
        ? 'clip'
        : 'hidden',
      lineHeight: '24px',
      '.ProseMirror': {
        margin: token('space.150', '12px'),
      },
      '.gridParent': {
        marginLeft: `${CommentEditorMargin - GRID_GUTTER}px`,
        marginRight: `${CommentEditorMargin - GRID_GUTTER}px`,
        width: `calc(100% + ${CommentEditorMargin - GRID_GUTTER}px)`,
      },
      padding: token('space.250', '20px'),
    },
    tableCommentEditorStyles,
  ),
);
ContentArea.displayName = 'ContentArea';

const secondaryToolbarStyle = css({
  boxSizing: 'border-box',
  justifyContent: 'flex-end',
  alignItems: 'center',
  display: 'flex',
  padding: `${token('space.150', '12px')} ${token('space.025', '2px')}`,
});
interface EditorAppearanceComponentState {}

const appearance: EditorAppearance = 'comment';

class Editor extends React.Component<
  EditorAppearanceComponentProps & WrappedComponentProps,
  EditorAppearanceComponentState
> {
  static displayName = 'CommentEditorAppearance';

  private appearance: EditorAppearance = 'comment';
  private containerElement: HTMLElement | null = null;

  // Wrapper container for toolbar and content area
  private wrapperElementRef = React.createRef<HTMLDivElement>();

  constructor(props: any) {
    super(props);
    if (props.innerRef) {
      this.wrapperElementRef = props.innerRef;
    }
  }

  private handleSave = () => {
    if (this.props.editorView && this.props.onSave) {
      this.props.onSave(this.props.editorView);
    }
  };

  private handleCancel = () => {
    if (this.props.editorView && this.props.onCancel) {
      this.props.onCancel(this.props.editorView);
    }
  };

  private renderChrome = ({ maxContentSize, mediaState }: PluginStates) => {
    const {
      editorDOMElement,
      editorView,
      editorActions,
      eventDispatcher,
      providerFactory,
      contentComponents,
      customContentComponents,
      customPrimaryToolbarComponents,
      primaryToolbarComponents,
      customSecondaryToolbarComponents,
      popupsMountPoint,
      popupsBoundariesElement,
      popupsScrollableElement,
      maxHeight,
      minHeight = 150,
      onSave,
      onCancel,
      disabled,
      dispatchAnalyticsEvent,
      intl,
      useStickyToolbar,
      pluginHooks,
      featureFlags,
    } = this.props;
    const maxContentSizeReached = Boolean(
      maxContentSize?.maxContentSizeReached,
    );
    const showSecondaryToolbar =
      !!onSave || !!onCancel || !!customSecondaryToolbarComponents;

    const isShortcutToFocusToolbar = (event: KeyboardEvent) => {
      //Alt + F9 to reach first element in this main toolbar
      return event.altKey && (event.key === 'F9' || event.keyCode === 120);
    };

    const isTwoLineToolbarEnabled =
      !!customPrimaryToolbarComponents && !!featureFlags?.twoLineEditorToolbar;

    const handleEscape = (event: KeyboardEvent) => {
      if (!editorView?.hasFocus()) {
        editorView?.focus();
      }
      event.preventDefault();
      event.stopPropagation();
    };

    return (
      <WithFlash animate={maxContentSizeReached}>
        <div
          css={[
            commentEditorStyle,
            css({
              minHeight: `${minHeight}px`,
            }),
          ]}
          className="akEditor"
          ref={this.wrapperElementRef}
        >
          <MainToolbar
            useStickyToolbar={useStickyToolbar}
            twoLineEditorToolbar={isTwoLineToolbarEnabled}
          >
            <ToolbarArrowKeyNavigationProvider
              editorView={editorView}
              childComponentSelector={"[data-testid='ak-editor-main-toolbar']"}
              isShortcutToFocusToolbar={isShortcutToFocusToolbar}
              handleEscape={handleEscape}
              editorAppearance={this.appearance}
              useStickyToolbar={useStickyToolbar}
              intl={intl}
            >
              <Toolbar
                editorView={editorView!}
                editorActions={editorActions}
                eventDispatcher={eventDispatcher!}
                providerFactory={providerFactory!}
                appearance={this.appearance}
                items={primaryToolbarComponents}
                popupsMountPoint={popupsMountPoint}
                popupsBoundariesElement={popupsBoundariesElement}
                popupsScrollableElement={popupsScrollableElement}
                disabled={!!disabled}
                dispatchAnalyticsEvent={dispatchAnalyticsEvent}
                containerElement={this.containerElement}
                twoLineEditorToolbar={isTwoLineToolbarEnabled}
              />
              <div
                css={mainToolbarCustomComponentsSlotStyle(
                  isTwoLineToolbarEnabled,
                )}
              >
                {customPrimaryToolbarComponents}
              </div>
            </ToolbarArrowKeyNavigationProvider>
          </MainToolbar>
          <ClickAreaBlock editorView={editorView} editorDisabled={disabled}>
            <WidthConsumer>
              {({ width }) => {
                return (
                  <ContentArea
                    ref={(ref) => (this.containerElement = ref)}
                    css={
                      maxHeight
                        ? css({
                            maxHeight: `${maxHeight}px`,
                          })
                        : null
                    }
                    className={classnames('ak-editor-content-area', {
                      'less-margin': width < akEditorMobileBreakoutPoint,
                    })}
                    featureFlags={featureFlags}
                  >
                    {customContentComponents &&
                    'before' in customContentComponents
                      ? customContentComponents.before
                      : customContentComponents}
                    <PluginSlot
                      editorView={editorView}
                      editorActions={editorActions}
                      eventDispatcher={eventDispatcher}
                      dispatchAnalyticsEvent={dispatchAnalyticsEvent}
                      providerFactory={providerFactory}
                      appearance={this.appearance}
                      items={contentComponents}
                      popupsMountPoint={popupsMountPoint}
                      popupsBoundariesElement={popupsBoundariesElement}
                      popupsScrollableElement={popupsScrollableElement}
                      containerElement={this.containerElement}
                      disabled={!!disabled}
                      wrapperElement={this.wrapperElementRef.current}
                      pluginHooks={pluginHooks}
                    />
                    {editorDOMElement}
                    {customContentComponents &&
                    'after' in customContentComponents
                      ? customContentComponents.after
                      : null}
                  </ContentArea>
                );
              }}
            </WidthConsumer>
          </ClickAreaBlock>
        </div>
        {showSecondaryToolbar && (
          <div
            css={secondaryToolbarStyle}
            data-testid="ak-editor-secondary-toolbar"
          >
            <ButtonGroup>
              {!!onSave && (
                <Button
                  appearance="primary"
                  onClick={this.handleSave}
                  testId="comment-save-button"
                  isDisabled={
                    disabled || (mediaState && !mediaState.allUploadsFinished)
                  }
                >
                  {intl.formatMessage(messages.saveButton)}
                </Button>
              )}
              {!!onCancel && (
                <Button
                  appearance="subtle"
                  onClick={this.handleCancel}
                  isDisabled={disabled}
                >
                  {intl.formatMessage(messages.cancelButton)}
                </Button>
              )}
            </ButtonGroup>
            <span style={{ flexGrow: 1 }} />
            {customSecondaryToolbarComponents}
          </div>
        )}
      </WithFlash>
    );
  };

  render() {
    return <RenderWithPluginState renderChrome={this.renderChrome} />;
  }
}

interface PluginStates {
  maxContentSize?: MaxContentSizePluginState;
  mediaState?: MediaPluginState;
}

interface RenderChromeProps {
  renderChrome: (props: PluginStates) => React.ReactNode;
}

function RenderWithPluginState({ renderChrome }: RenderChromeProps) {
  const api =
    usePresetContext<
      [OptionalPlugin<MediaPlugin>, OptionalPlugin<MaxContentSizePlugin>]
    >();
  const { mediaState, maxContentSizeState } = useSharedPluginState(api, [
    'media',
    'maxContentSize',
  ]);

  return (
    <Fragment>
      {renderChrome({
        maxContentSize: maxContentSizeState,
        mediaState: mediaState ?? undefined,
      })}
    </Fragment>
  );
}

const EditorNext = (
  props: EditorAppearanceComponentProps & WrappedComponentProps,
) => {
  const api =
    usePresetContext<
      [OptionalPlugin<MediaPlugin>, OptionalPlugin<MaxContentSizePlugin>]
    >();
  const { mediaState, maxContentSizeState } = useSharedPluginState(api, [
    'media',
    'maxContentSize',
  ]);
  const {
    editorDOMElement,
    editorView,
    editorActions,
    eventDispatcher,
    providerFactory,
    contentComponents,
    customContentComponents,
    customPrimaryToolbarComponents,
    primaryToolbarComponents,
    customSecondaryToolbarComponents,
    popupsMountPoint,
    popupsBoundariesElement,
    popupsScrollableElement,
    maxHeight,
    minHeight = 150,
    onSave,
    onCancel,
    disabled,
    dispatchAnalyticsEvent,
    intl,
    useStickyToolbar,
    pluginHooks,
    featureFlags,
    innerRef,
  } = props;
  const maxContentSizeReached = Boolean(
    maxContentSizeState?.maxContentSizeReached,
  );
  const showSecondaryToolbar =
    !!onSave || !!onCancel || !!customSecondaryToolbarComponents;
  let containerElement: HTMLElement | null = null;

  // Wrapper container for toolbar and content area
  const wrapperElementRef = useMemo(
    () => innerRef || React.createRef<HTMLDivElement>(),
    [innerRef],
  );

  const [saveButtonDisabled, setSaveButtonDisabled] = useState(disabled);

  useEffect(() => {
    if (mediaState) {
      mediaState.subscribeToUploadInProgressState(setSaveButtonDisabled);
    }
    return () =>
      mediaState?.unsubscribeFromUploadInProgressState(setSaveButtonDisabled);
  }, [mediaState]);

  const handleSave = useCallback(() => {
    if (editorView && onSave) {
      onSave(editorView);
    }
  }, [editorView, onSave]);

  const handleCancel = useCallback(() => {
    if (editorView && onCancel) {
      onCancel(editorView);
    }
  }, [editorView, onCancel]);

  const isShortcutToFocusToolbar = useCallback((event: KeyboardEvent) => {
    //Alt + F9 to reach first element in this main toolbar
    return event.altKey && (event.key === 'F9' || event.keyCode === 120);
  }, []);

  const isTwoLineToolbarEnabled =
    !!customPrimaryToolbarComponents && !!featureFlags?.twoLineEditorToolbar;

  const handleEscape = useCallback(
    (event: KeyboardEvent) => {
      if (!editorView?.hasFocus()) {
        editorView?.focus();
      }
      event.preventDefault();
      event.stopPropagation();
    },
    [editorView],
  );

  return (
    <WithFlash animate={maxContentSizeReached}>
      <div
        css={[
          commentEditorStyle,
          css({
            minHeight: `${minHeight}px`,
          }),
        ]}
        className="akEditor"
        ref={wrapperElementRef}
      >
        <MainToolbar
          useStickyToolbar={useStickyToolbar}
          twoLineEditorToolbar={isTwoLineToolbarEnabled}
        >
          <ToolbarArrowKeyNavigationProvider
            editorView={editorView}
            childComponentSelector={"[data-testid='ak-editor-main-toolbar']"}
            isShortcutToFocusToolbar={isShortcutToFocusToolbar}
            handleEscape={handleEscape}
            editorAppearance={appearance}
            useStickyToolbar={useStickyToolbar}
            intl={intl}
          >
            <Toolbar
              editorView={editorView!}
              editorActions={editorActions}
              eventDispatcher={eventDispatcher!}
              providerFactory={providerFactory!}
              appearance={appearance}
              items={primaryToolbarComponents}
              popupsMountPoint={popupsMountPoint}
              popupsBoundariesElement={popupsBoundariesElement}
              popupsScrollableElement={popupsScrollableElement}
              disabled={!!disabled}
              dispatchAnalyticsEvent={dispatchAnalyticsEvent}
              containerElement={containerElement}
              twoLineEditorToolbar={isTwoLineToolbarEnabled}
            />
            <div
              css={mainToolbarCustomComponentsSlotStyle(
                isTwoLineToolbarEnabled,
              )}
            >
              {customPrimaryToolbarComponents}
            </div>
          </ToolbarArrowKeyNavigationProvider>
        </MainToolbar>
        <ClickAreaBlock editorView={editorView} editorDisabled={disabled}>
          <WidthConsumer>
            {({ width }) => {
              return (
                <ContentArea
                  ref={(ref) => (containerElement = ref)}
                  css={
                    maxHeight
                      ? css({
                          maxHeight: `${maxHeight}px`,
                        })
                      : null
                  }
                  className={classnames('ak-editor-content-area', {
                    'less-margin': width < akEditorMobileBreakoutPoint,
                  })}
                  featureFlags={featureFlags}
                >
                  {customContentComponents &&
                  'before' in customContentComponents
                    ? customContentComponents.before
                    : customContentComponents}
                  <PluginSlot
                    editorView={editorView}
                    editorActions={editorActions}
                    eventDispatcher={eventDispatcher}
                    dispatchAnalyticsEvent={dispatchAnalyticsEvent}
                    providerFactory={providerFactory}
                    appearance={appearance}
                    items={contentComponents}
                    popupsMountPoint={popupsMountPoint}
                    popupsBoundariesElement={popupsBoundariesElement}
                    popupsScrollableElement={popupsScrollableElement}
                    containerElement={containerElement}
                    disabled={!!disabled}
                    wrapperElement={wrapperElementRef.current}
                    pluginHooks={pluginHooks}
                  />
                  {editorDOMElement}
                  {customContentComponents && 'after' in customContentComponents
                    ? customContentComponents.after
                    : null}
                </ContentArea>
              );
            }}
          </WidthConsumer>
        </ClickAreaBlock>
      </div>
      {showSecondaryToolbar && (
        <div
          css={secondaryToolbarStyle}
          data-testid="ak-editor-secondary-toolbar"
        >
          <ButtonGroup>
            {!!onSave && (
              <Button
                appearance="primary"
                onClick={handleSave}
                testId="comment-save-button"
                isDisabled={disabled || saveButtonDisabled}
              >
                {intl.formatMessage(messages.saveButton)}
              </Button>
            )}
            {!!onCancel && (
              <Button
                appearance="subtle"
                onClick={handleCancel}
                isDisabled={disabled}
              >
                {intl.formatMessage(messages.cancelButton)}
              </Button>
            )}
          </ButtonGroup>
          <span style={{ flexGrow: 1 }} />
          {customSecondaryToolbarComponents}
        </div>
      )}
    </WithFlash>
  );
};

EditorNext.displayName = 'CommentEditorAppearance';

const CommentEditorNextWithIntl = injectIntl(EditorNext);
const CommentEditorOldWithIntl = injectIntl(Editor);

const ExportComp = getBooleanFF(
  'platform.editor.media.alluploadsfinished-dispatch-update_ivtow',
)
  ? CommentEditorNextWithIntl
  : CommentEditorOldWithIntl;

export const CommentEditorWithIntl = ExportComp;
