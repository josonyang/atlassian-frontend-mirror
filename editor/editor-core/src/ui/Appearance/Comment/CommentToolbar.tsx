import React from 'react';

import { TOOLBARS } from '@atlaskit/editor-common/toolbar';
import type {
	EditorAppearance,
	OptionalPlugin,
	PublicPluginAPI,
} from '@atlaskit/editor-common/types';
import type { ToolbarPlugin } from '@atlaskit/editor-plugins/toolbar';
import type { EditorView } from '@atlaskit/editor-prosemirror/view';

import { isToolbar } from '../../../utils/toolbar';
import { ToolbarNext } from '../../Toolbar/Toolbar';

type ToolbarEditorPlugins = [OptionalPlugin<ToolbarPlugin>];

type CommentToolbarProps = {
	editorAPI?: PublicPluginAPI<ToolbarEditorPlugins>;
	editorAppearance: EditorAppearance;
	editorView?: EditorView;
};

/**
 * Primary toolbar driven by components registered by `editor-plugin-toolbar`, introduced in `platform_editor_toolbar_aifc`.
 */
export const CommentToolbar = ({
	editorAPI,
	editorView,
	editorAppearance,
}: CommentToolbarProps) => {
	const components = editorAPI?.toolbar?.actions.getComponents();
	const toolbar = components?.find((component) => component.key === TOOLBARS.PRIMARY_TOOLBAR);

	if (!components || !isToolbar(toolbar)) {
		return null;
	}

	return (
		<ToolbarNext
			toolbar={toolbar}
			components={components}
			editorView={editorView}
			editorAPI={editorAPI}
			editorAppearance={editorAppearance}
		/>
	);
};
