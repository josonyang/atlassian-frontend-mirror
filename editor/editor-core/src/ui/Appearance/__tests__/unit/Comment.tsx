import React from 'react';

import type { ReactWrapper } from 'enzyme';

import { ProviderFactory } from '@atlaskit/editor-common/provider-factory';
import type { DocBuilder } from '@atlaskit/editor-common/types';
import type { MediaOptions } from '@atlaskit/editor-plugins/media/types';
// eslint-disable-next-line import/no-extraneous-dependencies -- Removed import for fixing circular dependencies
import { createEditorFactory } from '@atlaskit/editor-test-helpers/create-editor';
// eslint-disable-next-line import/no-extraneous-dependencies -- Removed import for fixing circular dependencies
import { doc, p } from '@atlaskit/editor-test-helpers/doc-builder';
// eslint-disable-next-line import/no-extraneous-dependencies -- Removed import for fixing circular dependencies
import { sleep } from '@atlaskit/editor-test-helpers/sleep';
import { getDefaultMediaClientConfig } from '@atlaskit/media-test-helpers/fakeMediaClient';

import { mountWithIntl } from '../../../../__tests__/__helpers/enzyme';
import EditorActions from '../../../../actions';
import EditorContext from '../../../EditorContext';
import Comment from '../../Comment';

describe('comment editor', () => {
	afterEach(() => {
		jest.clearAllMocks();
	});

	const createEditor = createEditorFactory();

	const editor = (doc: DocBuilder) =>
		createEditor({
			doc,
			editorProps: { allowExtension: true },
		});
	it('should create empty terminal empty paragraph when clicked outside editor', () => {
		const { editorView } = editor(doc(p('Hello world'), p('Hello world')));
		const fullPage = mountWithIntl(
			<Comment
				editorAPI={undefined}
				editorView={editorView}
				providerFactory={{} as any}
				editorDOMElement={<div />}
				featureFlags={{}}
			/>,
		);

		fullPage.find('div[data-testid="click-wrapper"]').simulate('mousedown', { clientY: 200 });
		expect(editorView.state.doc).toEqualDocument(doc(p('Hello world'), p('Hello world'), p('')));
	});

	it('should not create empty terminal empty paragraph if it is already present at end', () => {
		const { editorView } = editor(doc(p('Hello world'), p('')));
		const fullPage = mountWithIntl(
			<Comment
				editorAPI={undefined}
				editorView={editorView}
				providerFactory={{} as any}
				editorDOMElement={<div />}
				featureFlags={{}}
			/>,
		);
		fullPage
			.find('div[data-testid="click-wrapper"]')
			.simulate('click', { clientY: 200 })
			.simulate('click', { clientY: 200 });
		expect(editorView.state.doc).toEqualDocument(doc(p('Hello world'), p('')));
	});

	it('should not create empty terminal paragraph when clicked inside editor', () => {
		const { editorView } = editor(doc(p('Hello world')));
		const fullPage = mountWithIntl(
			<Comment
				editorAPI={undefined}
				editorView={editorView}
				providerFactory={{} as any}
				editorDOMElement={<div />}
				featureFlags={{}}
			/>,
		);

		fullPage.find('div.ak-editor-content-area').simulate('click');

		expect(editorView.state.doc).toEqualDocument(doc(p('Hello world')));
	});

	describe('with media', () => {
		const mediaProvider = Promise.resolve({
			viewMediaClientConfig: getDefaultMediaClientConfig(),
		});
		const providerFactory = ProviderFactory.create({
			mediaProvider,
		});

		function getSaveButton(wrapper: ReactWrapper<any, any>) {
			return wrapper.find('button[data-testid="comment-save-button"]');
		}

		it('should not be disabled when mediaPluginState.allowUploadFinished is false', async () => {
			const { editorView, eventDispatcher, editorAPI } = createEditor({
				doc: doc(p('')),
				providerFactory,
				editorProps: {
					allowExtension: true,
					media: { allowMediaSingle: true },
					appearance: 'comment',
				},
			});

			const comment = mountWithIntl(
				<EditorContext editorActions={EditorActions.from(editorView, eventDispatcher)}>
					<Comment
						editorAPI={undefined}
						onSave={jest.fn()}
						editorView={editorView}
						providerFactory={providerFactory}
						editorDOMElement={<div />}
						featureFlags={{}}
					/>
				</EditorContext>,
			);
			const mediaPluginState = editorAPI?.media?.sharedState.currentState();

			mediaPluginState?.updateAndDispatch({
				allUploadsFinished: false,
			});

			await sleep(0);

			mediaPluginState?.updateAndDispatch({
				allUploadsFinished: true,
			});
			await sleep(0);

			expect(getSaveButton(comment).prop('disabled')).toBe(false);
		});

		it('should not be remain disabled when disabled prop has been updated', async () => {
			const { editorView, eventDispatcher } = createEditor({
				doc: doc(p('')),
				providerFactory,
				editorProps: {
					allowExtension: true,
					media: { allowMediaSingle: true },
					appearance: 'comment',
				},
			});

			const EditorComment = ({ disabled = true }) => (
				<EditorContext editorActions={EditorActions.from(editorView, eventDispatcher)}>
					<Comment
						editorAPI={undefined}
						onSave={jest.fn()}
						editorView={editorView}
						providerFactory={providerFactory}
						editorDOMElement={<div />}
						featureFlags={{}}
						disabled={disabled}
					/>
				</EditorContext>
			);
			const comment = mountWithIntl(<EditorComment disabled={true} />);
			await comment.setProps({ disabled: false });

			// save button should not be disabled
			expect(getSaveButton(comment).prop('disabled')).toBe(false);
		});

		it('should set up required media options for Comment Editor', () => {
			const { editorAPI } = createEditor({
				doc: doc(p('')),
				providerFactory,
				editorProps: {
					allowExtension: true,
					media: { allowMediaSingle: true },
					appearance: 'comment',
				},
			});

			const mediaPluginState = editorAPI?.media?.sharedState.currentState();
			expect(mediaPluginState?.mediaOptions).toBeDefined();

			const { allowAdvancedToolBarOptions, alignLeftOnInsert } =
				mediaPluginState?.mediaOptions as MediaOptions;
			expect(alignLeftOnInsert).toBe(true);
			expect(allowAdvancedToolBarOptions).toBe(true);
		});
	});
	describe('secondary toolbar', () => {
		it('should render the secondary toolbar if there is a save button', () => {
			const { editorView } = editor(doc(p('Hello world')));
			const fullPage = mountWithIntl(
				<Comment
					editorAPI={undefined}
					editorView={editorView}
					onSave={true as any}
					providerFactory={{} as any}
					editorDOMElement={<div />}
					featureFlags={{}}
				/>,
			);
			fullPage.find('div.ak-editor-content-area').simulate('click');
			expect(fullPage.find('div[data-testid="ak-editor-secondary-toolbar"]').exists()).toBe(true);
		});
		it('should render the secondary toolbar if there is a cancel button', () => {
			const { editorView } = editor(doc(p('Hello world')));
			const fullPage = mountWithIntl(
				<Comment
					editorAPI={undefined}
					editorView={editorView}
					onCancel={true as any}
					providerFactory={{} as any}
					editorDOMElement={<div />}
					featureFlags={{}}
				/>,
			);
			fullPage.find('div.ak-editor-content-area').simulate('click');

			expect(fullPage.find('div[data-testid="ak-editor-secondary-toolbar"]').exists()).toBe(true);
		});
		it('should render the secondary toolbar if there is a custom secondary toolbar button', () => {
			const { editorView } = editor(doc(p('Hello world')));
			const fullPage = mountWithIntl(
				<Comment
					editorAPI={undefined}
					editorView={editorView}
					customSecondaryToolbarComponents={true as any}
					providerFactory={{} as any}
					editorDOMElement={<div />}
					featureFlags={{}}
				/>,
			);
			fullPage.find('div.ak-editor-content-area').simulate('click');
			expect(fullPage.find('div[data-testid="ak-editor-secondary-toolbar"]').exists()).toBe(true);
		});
		it('should not render the secondary toolbar if there is no save, cancel or custom button', () => {
			const { editorView } = editor(doc(p('Hello world')));
			const fullPage = mountWithIntl(
				<Comment
					editorAPI={undefined}
					editorView={editorView}
					providerFactory={{} as any}
					editorDOMElement={<div />}
					featureFlags={{}}
				/>,
			);
			fullPage.find('div.ak-editor-content-area').simulate('click');
			expect(fullPage.find('div[data-testid="ak-editor-secondary-toolbar"]').exists()).toBe(false);
		});

		describe('comment toolbar shortcuts', () => {
			beforeAll(() => {
				// scrollIntoView is not available in jsdom so need to mock it https://github.com/jsdom/jsdom/issues/1695
				window.HTMLElement.prototype.scrollIntoView = jest.fn();
			});

			it('focuses editor on ESC', () => {
				const { editorView, commentComponent } = mountCommentWithToolbarButton();
				const editorFocusSpy = jest.spyOn(editorView, 'focus');

				const toolbarClickWrapper = commentComponent
					.find('.custom-key-handler-wrapper')
					.last()
					.getDOMNode();
				toolbarClickWrapper.dispatchEvent(
					new KeyboardEvent('keydown', {
						key: 'Escape',
					}),
				);
				expect(editorFocusSpy).toHaveBeenCalled();
			});

			it('focuses toolbar on alt + F9', () => {
				const { editorView, commentComponent } = mountCommentWithToolbarButton();
				const buttonElement = commentComponent
					.find('[data-testid="custom-button"]')
					.last()
					.getDOMNode() as HTMLElement;

				const buttonFocusSpy = jest.spyOn(buttonElement, 'focus');
				const buttonScrollSpy = jest.spyOn(buttonElement, 'scrollIntoView');

				editorView.dom.dispatchEvent(
					new KeyboardEvent('keydown', {
						key: 'F9',
						keyCode: 120,
						altKey: true,
					}),
				);
				expect(buttonFocusSpy).toHaveBeenCalled();
				expect(buttonScrollSpy).toHaveBeenCalled();
			});

			function mountCommentWithToolbarButton() {
				const { editorView } = editor(doc(p('Hello world')));
				const commentComponent = mountWithIntl(
					<Comment
						editorAPI={undefined}
						editorView={editorView}
						providerFactory={{} as any}
						editorDOMElement={<div />}
						primaryToolbarComponents={[() => <button data-testid="custom-button">Test</button>]}
						featureFlags={{}}
					/>,
				);
				return { editorView, commentComponent };
			}
		});
	});
});
