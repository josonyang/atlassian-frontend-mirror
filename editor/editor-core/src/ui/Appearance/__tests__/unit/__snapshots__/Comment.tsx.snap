// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`comment editor fixed toolbar styles platform-visual-refresh-icons is on visual refresh icons should render sticky toolbar with correct styles 1`] = `
".css-b7pr3x-FixedToolbar{position:relative;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;padding:var(--ds-space-100, 8px) var(--ds-space-100, 8px) 0;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;height:auto;background-color:var(--ds-surface, white);box-shadow:none;padding-left:var(--ds-space-250, 20px);}
.css-b7pr3x-FixedToolbar >div{}
.css-b7pr3x-FixedToolbar >div >div:first-of-type:not(style),.css-b7pr3x-FixedToolbar >div >style:first-of-type+*{margin-left:0;}
.css-b7pr3x-FixedToolbar .block-type-btn{padding-left:0;}
.css-161bjmo-FixedToolbar{position:relative;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;padding:var(--ds-space-100, 8px) var(--ds-space-100, 8px) 0;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;height:auto;background-color:var(--ds-surface, white);box-shadow:none;padding-left:var(--ds-space-250, 20px);}
.css-161bjmo-FixedToolbar >div{}
.css-161bjmo-FixedToolbar >div >div:first-of-type:not(style),.css-161bjmo-FixedToolbar >div >style:first-of-type+*{margin-left:0;}
.css-161bjmo-FixedToolbar .block-type-btn{padding-left:0;}
.css-161bjmo-FixedToolbar>div{}
@media (max-width: 490px){.css-161bjmo-FixedToolbar>div{-webkit-flex-direction:column-reverse;-ms-flex-direction:column-reverse;flex-direction:column-reverse;-webkit-align-items:end;-webkit-box-align:end;-ms-flex-align:end;align-items:end;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-box-pack:end;-ms-flex-pack:end;-webkit-justify-content:flex-end;justify-content:flex-end;}}
.css-161bjmo-FixedToolbar>div >div:first-of-type{}
@media (max-width: 490px){.css-161bjmo-FixedToolbar>div >div:first-of-type >div:nth-of-type(2){-webkit-box-pack:end;-ms-flex-pack:end;-webkit-justify-content:flex-end;justify-content:flex-end;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}}
.css-161bjmo-FixedToolbar span svg{max-width:100%;}"
`;

exports[`comment editor sticky toolbar styles platform-visual-refresh-icons is on visual refresh icons on should render sticky toolbar with correct styles 1`] = `
".css-qv9fzm-StickyToolbar{position:relative;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;padding:var(--ds-space-100, 8px) var(--ds-space-100, 8px) 0;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;height:auto;background-color:var(--ds-surface, white);box-shadow:none;padding-left:var(--ds-space-250, 20px);position:-webkit-sticky;position:sticky;padding-bottom:var(--ds-space-100, 8px);z-index:500;-webkit-transition:box-shadow ease-in-out 0.2s;transition:box-shadow ease-in-out 0.2s;}
.css-qv9fzm-StickyToolbar >div{}
.css-qv9fzm-StickyToolbar >div >div:first-of-type:not(style),.css-qv9fzm-StickyToolbar >div >style:first-of-type+*{margin-left:0;}
.css-qv9fzm-StickyToolbar .block-type-btn{padding-left:0;}
.css-qv9fzm-StickyToolbar>div{}
@media (max-width: 490px){.css-qv9fzm-StickyToolbar>div{-webkit-flex-direction:column-reverse;-ms-flex-direction:column-reverse;flex-direction:column-reverse;-webkit-align-items:end;-webkit-box-align:end;-ms-flex-align:end;align-items:end;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-box-pack:end;-ms-flex-pack:end;-webkit-justify-content:flex-end;justify-content:flex-end;}}
.css-qv9fzm-StickyToolbar>div >div:first-of-type{}
@media (max-width: 490px){.css-qv9fzm-StickyToolbar>div >div:first-of-type >div:nth-of-type(2){-webkit-box-pack:end;-ms-flex-pack:end;-webkit-justify-content:flex-end;justify-content:flex-end;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}}
.css-qv9fzm-StickyToolbar span svg{max-width:100%;}
.css-qv9fzm-StickyToolbar.show-keyline{box-shadow:0 2px 0 0 var(--ds-background-accent-gray-subtlest, #F1F2F4);}"
`;
